### PR TITLE
[1004] Wrap service loader calls in privileged actions if the securit…

### LINKED ISSUE #1004: Partial implementation of privileged actions should be completed
In the various `FactoryFinder`'s there is a privileged action used to get the context class loader. However, there is no such privileged action around the service loader which could fail if the security manager is enabled and read access is not explicitly granted to the implementation JAR.

The `ClientBuilder` could also fail with a security issue while attempting to throw a `LinkageError` if the security manager is enabled and the correct permissions are not defined.

### PR DESCRIPTION
**Pull request by jamezp**
…y manager is enabled. Also ensure the LinkageError for the ClientBuilder is thrown if a security manager is enabled.

Resolves #1004

Signed-off-by: James R. Perkins <jperkins@redhat.com>

### CODE REVIEW

**Comment by jamezp** [comment]
I forgot to mention, if this seems acceptable I can port the changes over to whichever branches desired.

**Comment by arjantijms** [comment]
@jamezp do note that `SecurityManager`, `AccessController` etc  have been deprecated for removal in JDK 17. So we may want to remove such privilege actions, not add them.

**Comment by jansupol** [comment]
There is quite a lot of code duplication in the world of suppliers.

[JEP 411](https://openjdk.java.net/jeps/411) deprecates the SecurityManager in JDK 17, but for JDK 17, it means a warning is issued. Jakarta REST 3.1 is supposed to be working with JDK 11-17 which still has the SecurityManager functional. But I'm surprised noone complained about this untill now (2.0 was released with EE 7) 

@jamezp You seem to use the words may/might, is this a real issue or the could-potentially-happen one?

**Comment by jamezp** [comment]
> @jamezp do note that `SecurityManager`, `AccessController` etc have been deprecated for removal in JDK 17. So we may want to remove such privilege actions, not add them.

Yes, I thought about that before sending this. However, there still is no clear path for a replacement that I'm aware of. I know that some installations require the security manager to be enabled.

The real purpose of this is so RESTEasy no longer has to use a fork of the spec :)

**Comment by jamezp** [comment]
> There is quite a lot of code duplication in the world of suppliers.
> 
> [JEP 411](https://openjdk.java.net/jeps/411) deprecates the SecurityManager in JDK 17, but for JDK 17, it means a warning is issued. Jakarta REST 3.1 is supposed to be working with JDK 11-17 which still has the SecurityManager functional. But I'm surprised noone complained about this untill now (2.0 was released with EE 7)

Yes. AFAIK there is no replacement for the security manager at this point. I have a feeling it won't be removed anytime soon :)

> 
> @jamezp You seem to use the words may/might, is this a real issue or the could-potentially-happen one?

Sorry, yes this is a real issue. I've not tried in other application servers, but I can say in WildFly if you run with a security manager then there are failures. There is a fix in a [fork](https://github.com/jboss/jboss-jakarta-jaxrs-api_spec) which WildFly and JBoss EAP use. However it would be nice to let the fork die and use the spec itself :)

**Comment by mkarg** [comment]
I'd rather vote for removal of security manager instead of correcting it.

**Comment by spericas** [comment]
@jamezp What is the use case for running WildFly under a security manager? The JDK team seems to simply by JEP 411 that this is no longer needed/used in server-side Java.

**Comment by andymc12** [comment]
I know that we (IBM) have customers who require Java 2 security for their app servers - even in Java/Jakarta EE.  Personally, I've never been a fan of Java 2 security, but we do have some users who need it.  I believe most of our Java 2 security customers are governmental users or are contractors for governments that require it. Those requirements might start to be lifted now that Java 17 is deprecating the security manager, but in general (and until we get a little closer to prereq'ing Java 17), I'm in favor of adding doPrivs where necessary.

**Comment by jamezp** [comment]
> @jamezp What is the use case for running WildFly under a security manager? The JDK team seems to simply by JEP 411 that this is no longer needed/used in server-side Java.

@spericas I can't say I know the requirements specifically for WildFly, but I know we (Red Hat) have JBoss EAP customers that are required to run the container with the security manager enabled.

I agree with @andymc12 that I don't like it, but I don't really have a choice but to support it :)

File: jaxrs-api/src/main/java/jakarta/ws/rs/client/ClientBuilder.java [hunk 1]
```diff
@@ -59,16 +61,31 @@ public static ClientBuilder newBuilder() {
...
+                    if (loader == null) {
+                        loader = ClassLoader.getSystemClassLoader();
+                    }
+                    URL targetTypeURL = loader.getResource(classnameAsResource);
+                    errorMessage = "ClassCastException: attempting to cast"
+                            + delegate.getClass().getClassLoader().getResource(classnameAsResource)
+                            + " to " + targetTypeURL;
+                } else {
+                    errorMessage = AccessController.doPrivileged((PrivilegedAction<String>) () -> {
+                        ClassLoader loader = pClass.getClassLoader();
```
Comment (mkarg) [inline jaxrs-api/src/main/java/jakarta/ws/rs/client/ClientBuilder.java:78]: Maybe we should refactor into private method instead of duplicating code.

**Comment by spericas** [comment]
@jamezp @andymc12 Just reviewed JEP 411. Looks to me that it would be OK to fix this for 3.1. We should probably review this for 4.0 as clearly a lot of this functionality is going away. I still think that the "security" provided by the Java Security Manager is a lot less useful in practice than on paper.

File: jaxrs-api/src/main/java/jakarta/ws/rs/client/ClientBuilder.java (hunk 1 above)
Comment (jamezp) [reply jaxrs-api/src/main/java/jakarta/ws/rs/client/ClientBuilder.java]: Should be done if I've understood correctly :)

**Review by spericas — DISMISSED** [review]
LGTM. Only comment is that `findFirstService`'s could create an `Action` to avoid duplicating code blocks.

File: jaxrs-api/src/main/java/jakarta/ws/rs/client/FactoryFinder.java [hunk 2]
```diff
@@ -168,4 +163,44 @@ private static Object newInstance(final String className, final ClassLoader clas
...
+        }
+        return AccessController.doPrivileged((PrivilegedAction<ClassLoader>) FactoryFinder.class::getClassLoader);
+    }
+
+    private static <T> T findFirstService(final String factoryId, final ClassLoader cl, final Class<T> service) {
+        final FindServiceAction<T> action = new FindServiceAction<>(factoryId, cl, service);
+        if (System.getSecurityManager() == null) {
+            return action.run();
+        }
+        return AccessController.doPrivileged(action);
```
Comment (andymc12) [inline jaxrs-api/src/main/java/jakarta/ws/rs/client/FactoryFinder.java:179]: I understand putting the `getClassLoader` calls into a doPriv block, but I don't understand the `ServiceLoader` calls.  Does that need to be in a doPriv?  The javadoc doesn't mention that it throws a `SecurityException`.

File: jaxrs-api/src/main/java/jakarta/ws/rs/ext/FactoryFinder.java [hunk 3]
```diff
@@ -168,4 +163,44 @@ private static Object newInstance(final String className, final ClassLoader clas
...
+        }
+        return AccessController.doPrivileged((PrivilegedAction<ClassLoader>) FactoryFinder.class::getClassLoader);
+    }
+
+    private static <T> T findFirstService(final String factoryId, final ClassLoader cl, final Class<T> service) {
+        final FindServiceAction<T> action = new FactoryFinder.FindServiceAction<>(factoryId, cl, service);
+        if (System.getSecurityManager() == null) {
+            return action.run();
+        }
+        return AccessController.doPrivileged(action);
```
Comment (andymc12) [inline jaxrs-api/src/main/java/jakarta/ws/rs/ext/FactoryFinder.java:179]: Same comment as above. Can we avoid the doPriv on the `ServiceLoader.load(...)` calls?

**Review by andymc12 — COMMENTED** [review]
Overall, I like this changes - the performance changes and adding `getClassLoader` calls into doPriv blocks.  I only have a question about whether `ServiceLoader.load(...)` and subsequent iterating requires a doPriv block.  Thanks!

File: jaxrs-api/src/main/java/jakarta/ws/rs/client/FactoryFinder.java (hunk 2 above)
Comment (jamezp) [reply jaxrs-api/src/main/java/jakarta/ws/rs/client/FactoryFinder.java:179]: The `ServiceLoader.load()` is actually safe, but it captures the `AccessContrelContext` is captured and during iteration a security exception could be thrown on `Constructor.newInstance()`.

From the JavaDocs:

> Service loaders always execute in the security context of the caller. Trusted system code should typically invoke the methods in this class, and the methods of the iterators which they return, from within a privileged security context.

**Comment by spericas** [comment]
> > LGTM. Only comment is that `findFirstService`'s could create an `Action` to avoid duplicating code blocks.
> 
> Done.

Note that `PrivilegedAction` is a functional interface, so I think you could simplify this using capturing lambdas for even better readability without the need inner classes.

**Comment by jamezp** [comment]
> Note that `PrivilegedAction` is a functional interface, so I think you could simplify this using capturing lambdas for even better readability without the need inner classes.

Good point. Updated!

**Review by spericas — APPROVED** [review]

**Comment by spericas** [comment]
Merging as it has been approved 25 days ago.

**Pull request merged**